Flash read

sjg20 opened the pull request:
Little enhancement to 'flash read'

sjg20 committed:
shell: Export shell_hexdump_line()

This function is useful for other code. Export it along with the
default line length in bytes.

Signed-off-by: Simon Glass <sjg@chromium.org>

sjg20 committed:
flash: Enhance 'flash read' to read a block of data

It is annoying to read just a single word at a time. Update this
command to print any amount of data. This uses byte format at present.

We could perhaps support something like:

   flash read.8
   flash read.16
   flash read.32

to chose the size.

Example output (with line breaks to keep within git-style limit):

$ flash read FLASH_CTRL 0 20
00000000: 20 25 00 20 1d 26 00 08  69 68 00 08 45 26 00 08
   | %. .&.. ih..E&..|
00000010: 00 00 00 00 00 00 00 00  00 00 00 00 00 00 00 00
   |........ ........|

Signed-off-by: Simon Glass <sjg@chromium.org>

nashif commented on include/util.h:
```diff
@@ -0,0 +1,22 @@
+#define ZEPHYR_INCLUDE_MATH_H_
+
+/** Return the maximum of a and b */
+#define max(a, b)	\
```
did you check MIN and MAX in include/sys/util.h?

sglass68 replied:
:-) OK I need to check my ctags settings. I thought it was odd.

nordic-krch commented on drivers/flash/flash_shell.c:
```diff
@@ -153,12 +156,14 @@ static int cmd_read(const struct shell *shell, size_t argc, char *argv[])
-		flash_read(flash_dev, addr, &data, sizeof(data));
-		shell_print(shell, "0x%08x ", data);
-		addr += sizeof(data);
+		todo = MIN(cnt, SHELL_HEXDUMP_BYTES_IN_LINE);
```
i think it should be `cnt - upto`

sglass68 replied:
oops missed this. Fixed now.

What is the best way to write a test for this? I could use the dummy shell backend perhaps?

nordic-krch replied:
not sure how flash works on qemu target and if that can be used for testing. @nvlsianpu ?

sglass68 replied:
Eek I don't want to use qemu. I just wondered if there was a native/posix driver.

nvlsianpu replied:
It does work on qemu an posix using flash simulator driver.

sglass68 replied:
Qemu is a large overhead for testing a simple thing. It looks like I can use posix for this, although I do see what are apparently bugs in the driver.

https://github.com/zephyrproject-rtos/zephyr/issues/25109

sglass68 replied:
I found a way to do a native test

sglass68 commented:
Fixed conflict

sjg20 committed:
tests: shell: shell_flash: Add a test for 'flash read'

Add a simple test for this command to check that the output is as
expected.

This requires a larger output buffer in shell_dummy, so update that
too.

Signed-off-by: Simon Glass <sjg@chromium.org>

sjg20 committed:
drivers: flash: flash_simulator: Fix 'succedd' typo

Fix the 'succedd' typo which is in lots of places in this file.

Signed-off-by: Simon Glass <sjg@chromium.org>

nordic-krch commented:
thanks for the tests. I would like to approve it. You just need to figure out what is wrong with CI. Many warning.

sglass68 commented:
> thanks for the tests. I would like to approve it. You just need to figure out what is wrong with CI. Many warning.

Yes, it needed a board filter. I figured out how to get the logs and how to run locally. However it takes 20 minutes which seems like a long time.

vanwinkeljan commented on drivers/flash/flash_simulator.c:
```diff
@@ -397,10 +397,8 @@ static int flash_mock_init(struct device *dev)
-		/* erase the memory unit by pulling all bits to one */
-		(void)memset(mock_flash, 0xff, FLASH_SIMULATOR_FLASH_SIZE);
-	}
+	/* erase the memory unit by pulling all bits to one */
```
This breaks persistency of the flash device, the flash should only be set to all 0xFF if not already exists.

sglass68 replied:
OK I dropped that commit

vanwinkeljan reviewed (dismissed):
flash persistency is a feature of native posix and flash device should not be erased every time

sjg20 committed:
drivers: flash: flash_simulator: Erase flash first

With native_posix, a flash.bin file is created when running the tests.
If the tests are re-run, they fail since the flash is no-longer erased.

Before running the tests, erase the flash so that the 'read' test can
work regardless of the previous state.

Fixes: 25171
Signed-off-by: Simon Glass <sjg@chromium.org>

sjg20 commented:
I seem to need more than two approvers for this one. Any takers?

MaureenHelm reviewed (approved):
@jukka why did you cancel the shippable run?

carlescufi commented:
@sjg20 please rebase
